### PR TITLE
[SoundCloud] Detect whether there are any more search results

### PR DESCRIPTION
**Pull request by TobiGr**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

Previously, the extractor always assumed that there was an infinite number of search results. That has caused some problems. When searching for something with only a few results, that could cause clients such as NewPipe to try to load more results without getting more and thus causing tons of requests, ultimately resulting in reCaptcha requests.

### CODE REVIEW

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudSearchExtractor.java [hunk 1]
```diff
@@ -74,17 +82,29 @@ public InfoItemsPage<InfoItem> getPage(final Page page) throws IOException,
...
+            totalResults = result.getInt(TOTAL_RESULTS);
         } catch (final JsonParserException e) {
             throw new ParsingException("Could not parse json response", e);
         }
+        final boolean hasNextPage;
+        try {
+            hasNextPage = getOffsetFromUrl(page.getUrl()) + ITEMS_PER_PAGE <= totalResults;
+        } catch (MalformedURLException | UnsupportedEncodingException e) {
+            throw new ParsingException("Could not get offset from page URL", e);
+        }
```
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudSearchExtractor.java:100]: Move the try-catch to the method itself

**Review by Stypox — CHANGES_REQUESTED** [review]

**Comment by Stypox** [comment]
@TobiGr I pushed 29472571110d1074f4c08af87d13f1db6ac8e950 and changed the `<=` into a `<`. There are more results only if the results fetched so far are **less** than the total amount, not less or equal. Just to make sure, I tested with ITEMS_PER_PAGE=1 and it works. However, the only downside of `<=` was an additional network request I think, I didn't run into infinite calls, so the code seems resilient. Thanks!

**Review by Stypox — APPROVED** [review]

**Review by AudricV — COMMENTED** [review]
A review that I didn't have the time to finish and submit.

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java [hunk 2]
```diff
@@ -72,8 +63,7 @@ public InfoItemsPage<CommentsInfoItem> getPage(final Page page) throws Extractio
                 getServiceId());
 
         collectStreamsFrom(collector, json.getArray("collection"));
-
-        return new InfoItemsPage<>(collector, new Page(json.getString("next_href")));
+        return new InfoItemsPage<>(collector, new Page(json.getString("next_href", null)));
```
[the remark this answers is not in the document]
Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudCommentsExtractor.java:66]: That's right, however this is fine, too

**Pull request merged**